File: 01Singleton/Students/2018/DebowskiKacper/Singleton/Singleton/ThreadSafeSingleton.swift
//
//  ThreadSafeSingleton.swift
//  Singleton
//
//  Created by Kacper Debowski on 07/12/2018.
//  Copyright © 2018 Coding Skies. All rights reserved.
//

import Foundation

class ThreadSafeSingleton: Codable {
    private static var instance: ThreadSafeSingleton? = nil
    
    var timestamp: Double = 0

    private init() {
        self.timestamp = NSDate().timeIntervalSince1970
    }
    
    static func getInstance() -> ThreadSafeSingleton {
        if instance == nil {
            objc_sync_enter(self)
            if instance == nil {
                usleep(100000)
                instance = ThreadSafeSingleton()
            }
            objc_sync_exit(self)
        }
        return instance!
    }
}

File: 01Singleton/Students/2019/AliaksandrNovik/src/main/java/threadsafe/ThreadSafeSingleton.java
package threadsafe;

public class ThreadSafeSingleton {

	private static ThreadSafeSingleton instance;

	private ThreadSafeSingleton() {}
	
	public static ThreadSafeSingleton getInstance() {
		if(instance==null) {
			synchronized (ThreadSafeSingleton.class) {
				instance = new ThreadSafeSingleton();
			}
		}
		return instance;
	}
	
	public long fibonacci(int n) {
		if (n <= 1) return n;
        else return fibonacci(n-1) + fibonacci(n-2);
	}

}
